debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java

```diff
@@ -0,0 +1,645 @@
+        }
+    }
+
+    /**
+     * Callback handler for the 'C' commit replication message.
+     *
+     * @param buffer The replication stream buffer
+     */
+    private void handleCommitMessage(ByteBuffer buffer) {
+        int flags = buffer.get(); // flags, currently unused
+        long lsn = buffer.getLong(); // LSN of the commit
+        long endLsn = buffer.getLong(); // End LSN of the transaction
+        Instant commitTimestamp = PG_EPOCH.plus(buffer.getLong(), ChronoUnit.MICROS);
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Event: {}", MessageType.COMMIT);
+            LOGGER.trace("Flags: {} (currently unused and most likely 0)", flags);
+            LOGGER.trace("Commit LSN: {}", lsn);
+            LOGGER.trace("End LSN of transaction: {}", endLsn);
+            LOGGER.trace("Commit timestamp of transaction: {}", commitTimestamp);
+        }
+    }
+
+    /**
+     * Callback handler for the 'R' relation replication message.
+     *
+     * @param buffer The replication stream buffer
+     * @param typeRegistry The postgres type registry
+     */
+    private void handleRelationMessage(ByteBuffer buffer, TypeRegistry typeRegistry) throws SQLException {
+        int relationId = buffer.getInt();
+        String namespace = readString(buffer);
+        String relationName = readString(buffer);
+        int replicaIdentityId = buffer.get();
+        short columnCount = buffer.getShort();
+
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Event: {}, RelationId: {}, Replica Identity: {}, Columns: {}", MessageType.RELATION, relationId, replicaIdentityId, columnCount);
+            LOGGER.trace("Schema: '{}', Table: '{}'", namespace, relationName);
+        }
+
+        List<ColumnMetaData> columns = new ArrayList<>();
+        for (short i = 0; i < columnCount; ++i) {
+            byte flags = buffer.get();
+            String columnName = Strings.unquoteIdentifierPart(readString(buffer));
+            int columnType = buffer.getInt();
+            int attypmod = buffer.getInt();
+
+            final PostgresType postgresType = typeRegistry.get(columnType);
+            boolean key = (flags & 1) == 1;
+
+            columns.add(new ColumnMetaData(columnName, postgresType, key, attypmod));
+        }
+
+        PgOutputRelationMetaData metadata = new PgOutputRelationMetaData(relationId, namespace, relationName, columns);
+        if (relationIdMetadata.containsKey(relationId)) {
+            LOGGER.warn("Prior relation message for '{}' has not been consumed by a DML; being replaced.", relationId);
+        }
+
+        TableId tableId = new TableId(null, namespace, relationName);
+
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Cached relation '{}' with table-id '{}' and synchronizing schema", relationId, tableId);
+        }
+
+        relationIdMetadata.put(relationId, metadata);
+        tableIdByRelationId.put(relationId, tableId);
```

Comment:
The pgoutput stream guarantees that for every new replication connection established, if a DML operation is being streamed, I'll get a relation message that describes the schema for that operation.  I won't get another relation message describing that relation-id schema again unless the DML operation being streamed reflects that the schema was modified.

My goal was to design the decoder to attempt to keep a low memory footprint.  Since we have a `PostgresSchema` that maintains the schema of a table and will be updated accordingly based on the incoming insert/update/delete messages, I needed a way to cache the relation messages temporarily until I could process at least one insert/update/delete operation to push the decoded schema changes into Debezium.  

So `relationIdMetadata` acts as a temporary cache, nothing more.  In theory, it should be empty once a replication connection has streamed all changes and is idle.

The `tableIdRelationId` map is there to provide long-running replication connection support for future insert/update/delete statements we get where we don't get a relation message because the schema hasn't changed; allowing us to reverse lookup a relation-id back to a `TableId`.

In discussion with Gunnar, he mentioned folding relation-id into `PostgresSchema`.  I looked at doing that the rabbit hole lead to quite a number of changes that felt wrong; perhaps it was my approach which was flawed.  Either way, since this is all pgoutput specific I decided for now it was simple enough to keep it encapsulated in the decoder until we could discuss this aspect in a bit more detail as a team.

With all that clarified, we could merge these two; however that would imply that we would then need to keep the `relationIdMetadata` indefinitely over the course of a replication connection, which I think may bring us full circle to the history topic?

Reply:
@Naros Good thinking with the memory. One last question, why can't you apply the relation directly to the schema when it arrives and you need to do it in two phase approach?

Reply:
It looks like that is exactly what I may be doing anyway, see the call the `synchronizeSchemaFromMetadata` inside method that handles the incoming relation method.  We may likely be able to remove this one map all together; good catch.  I'll try this out and get back to you if I run into any problems or push an update if it works.